### PR TITLE
Mobile friendly search

### PR DESCRIPTION
**Pull request by kulyk**
This PR improves search UI/UX when Metabase is used with a mobile device. Currently, there are a few inconveniences about how search works on mobile:

* the search bar takes the app bar space no matter if you're needing it right now
* the search bar is pretty narrow. However, when you're searching on mobile, it's clear it's your main action right now, so it deserves some more focus
* the sidebar toggle button remains visible and takes space when you're searching
* the search results are displayed in a narrow popover

The PR changes the mobile UX/UX, so now:

* the search input is not visible at all, but there is a search icon on the right side of the app bar
* once you tap the icon, the sidebar toggle button disappears and the search icon expands into a full-width input
* a list of recent items takes full width, so it's easier to read
* search results take full width and height, so they're easier to see
* once a search query is entered, a "clear" button appears to clean the input

### To Verify

1. Open any non-admin page, ensure the desktop search behaves the same way it does on `master`
2. Make the window narrow or use a mobile mode from the browser's developer tools
3. Ensure the search input is not visible, but there is a search icon on the right side of the sidebar
4. Click the icon. Make sure sidebar's toggle button disappears and the icon expands into a full width input (except the logo)
5. Ensure the recent items list takes full width
6. Enter some query, ensure a list of search results takes full width and height and you can scroll it if there are many results
7. Ensure you can click a recent item and a search results item and actually open them
8. Ensure you can click outside of the input to dismiss search and that everything turns back into the original state: visible sidebar toggle button, no search input, search icon

### Demo

**Before** 

https://user-images.githubusercontent.com/17258145/161558058-fd9a43da-9e43-430f-ac2d-a53ea2ca9538.mp4 

**After**

https://user-images.githubusercontent.com/17258145/161558068-9b605244-bf60-4192-bb7b-fd117bc7d594.mp4

### CODE REVIEW

**Review by gusaiani — COMMENTED** [review]
Would it be trivial to use this PR and align the top and bottom of the search box with the New button?

image

**Comment by kulyk** [comment]
> Would it be trivial to use this PR and align the top and bottom of the search box with the New button?

Yeah, it won't be too difficult. Wonder what @kdoh thinks

**Comment by gusaiani** [comment]
@mazameli @kdoh 

This PR made me think we might want to use this Magnifier icon even on desktop.
User clicks the icon, app focuses on the input. Same number of clicks/taps as before.
Could also use a keyboard shortcut to open the search box.

Would be so clean.

File: frontend/src/metabase/hooks/use-keyboard-shortcut.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { useEffect } from "react";
```
Comment (gusaiani) [inline frontend/src/metabase/hooks/use-keyboard-shortcut.ts:1]: Nice hook.

File: frontend/src/metabase/hooks/use-on-outside-click.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import { useEffect, RefObject } from "react";
+
+interface ValidRefTarget {
+  contains(target: EventTarget | null): boolean;
+}
+
+export function useOnOutsideClick<T extends ValidRefTarget = HTMLDivElement>(
```
Comment (gusaiani) [inline frontend/src/metabase/hooks/use-on-outside-click.ts:7]: Maybe consider calling it `useOnClickOutside`.

Would be inline with the function name in it, `handleClickOutside`.

Also seems more expected.

**Review by gusaiani — APPROVED** [review]
Much better.

One inlined comment.

Maybe a visual regression test for mobile might be useful.

**Review by kdoh — APPROVED** [review]
Looks good to me!

**Comment by kdoh** [comment]
@kulyk @gusaiani Yeah it'd be great to fix that alignment here if we can but open to a follow up as well.

**Comment by kulyk** [comment]
Going to merge this and open another PR to fix the heights. I'd need to clean up the `NewButton` component and a few other files for that, so let's better keep these changes separated

**Pull request merged**